### ISSUE TITLE
torch.einsum equation works in NumPy but not in Pytorch

### ISSUE THREAD

**Issue by stashvala** [pytorch/pytorch#15671]
## Issue description

I tried doing batchwise dot product across channels or rather pairwise similarity between all pairs of features for two sets of feature matrices using `torch.einsum`, but it fails. Could be a bug or is this to be expected?

## To Reproduce

```python
>>> import torch
>>> a = torch.randn(1, 3, 24, 20)
>>> b = torch.randn(5, 3, 24, 20)
>>> torch.einsum("bijk, bilm -> bjklm", a, b).shape
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.6/dist-packages/torch/functional.py", line 245, in einsum
    return torch._C._VariableFunctions.einsum(equation, operands)
RuntimeError: size of dimension does not match previous size, operand 1, dim 0
```

## Expected behavior

In NumPy i get
```python
>>> import numpy as np
>>> a = np.random.rand(1, 3, 24, 20)
>>> b = np.random.rand(5, 3, 24, 20)
>>> np.einsum("bijk, bilm -> bjklm", a, b).shape
(5, 24, 20, 24, 20)
```

## Environment

PyTorch version: 1.0.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 950M
Nvidia driver version: 396.44
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.1.4
/usr/lib/x86_64-linux-gnu/libcudnn_static_v7.a
/usr/local/cuda-9.2/lib64/libcudnn.so
/usr/local/cuda-9.2/lib64/libcudnn.so.7
/usr/local/cuda-9.2/lib64/libcudnn.so.7.1.4
/usr/local/cuda-9.2/lib64/libcudnn_static.a

Versions of relevant libraries:
[pip3] numpy (1.15.4)
[pip3] torch (1.0.0)
[pip3] torchvision (0.2.1)
[conda] Could not collect

cc @mruberry @rgommers @vincentqb @vishwakftw @jianyuh @nikitaved @pearu

**Comment by fmassa** [comment]
Cc @t-vi

**Comment by t-vi** [comment]
Yeah, I think we don't delete the whitespace yet. Should be easy enough to fix in PyTorch, but removing them from the equation is a good workaround until we do.

**Comment by vadimkantorov** [comment]
@t-vi Was it not fixed in https://github.com/pytorch/pytorch/pull/9994 ?

**Comment by stashvala** [comment]
@t-vi 
Any news? Can I bypass the error somehow? I would really benefit from the `einsum` speedup versus the regular approach with for loop and `bmm` or something similar.

**Comment by ezyang** [comment]
@stashvala Did deleting the spaces from the equation not work?

**Comment by t-vi** [comment]
@stashvala If you know which indices need broadcasting, you can add a dummy letter for them.
@ezyang No, it's a broadcasting thing upon closer look, sorry for the red herring re spaces.

It should be easy to fix and I'm adding it to my todo list.

**Comment by stashvala** [comment]
@t-vi Could you explain a bit more please? In my example of `torch.einsum("bijk, bilm -> bjklm", torch.randn(1, 3, 24, 20), torch.randn(5, 3, 24, 20))`, _b_ for batch is broadcastable. What char represents the dummy letter? Do you have any examples maybe? Thanks!

**Comment by t-vi** [comment]
Ah, sorry, I meant to post the workaround but forgot. Just replace the letter in size-1-dimension with anything not appearing in the equation, e.g.:
```python
torch.einsum("xijk, bilm -> bjklm", torch.randn(1, 3, 24, 20), torch.randn(5, 3, 24, 20))
```
This works because you can split out the constant (after broadcasting) from the product.

Best regards

Thomas

**Comment by stashvala** [comment]
@t-vi I get it now, what you meant by broadcastable dimension and dummy letter... It works, cool, thanks!

**Comment by xvdp** [comment]
This may or not be related, but I found no docs on the topic.
Ellipsis working in numpy are not working in torch einsum
Example:
```python
import numpy as np
import torch
np.__version__, torch.__version__
#('1.16.3', '1.1.0')
```
```python
M = np.random.randn(2,2,3)
N = np.random.randn(3,3)
np.einsum('...ij, ...j->...i', N, M)
#OK
```

```python
torch.einsum('...ij, ...j->...i', torch.from_numpy(N), torch.from_numpy(M))
#~/miniconda3/envs/abg3/lib/python3.7/site-packages/torch/functional.py in einsum(equation, *operands)
#    209         # the old interface of passing the operands as one list argument
#    210         operands = operands[0]
#--> 211     return torch._C._VariableFunctions.einsum(equation, operands)
# RuntimeError: ellipsis must represent 0 dimensions in all terms
```

**Comment by Balandat** [comment]
I'm seeing a similar issue as @xvdp, where, unlike in numpy leading dimensions of the first operator are not properly broadcasted. E.g:

```
dimsM = [11, 3, 2, 7, 5]
dimsN = [2, 5]

M = np.random.randn(*dimsM)
N = np.random.randn(*dimsN)

W_np = np.einsum('...ij, ...j->...i', N, M)  # works fine
```

```
M_t = torch.from_numpy(M)
N_t = torch.from_numpy(N)

W_t = torch.einsum('...ij, ...j->...i', [M_t, N_t])

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-55-338835e35070> in <module>()
----> 1 torch.einsum('...ij, ...j->...i', [M_t, N_t])

/data/users/balandat/fbsource/fbcode/buck-out/dev/gen/bento/kernels/bento_kernel_ae_lazarus#link-tree/torch/functional.py in einsum(equation, *operands)
    209         # the old interface of passing the operands as one list argument
    210         operands = operands[0]
--> 211     return torch._C._VariableFunctions.einsum(equation, operands)
    212 
    213 

RuntimeError: ellipsis must represent 3 dimensions in all terms
```

Here is a hacky fix that dynamically generates an formula with as many explicit leading letters as necessary:
```
import string

prefix = string.ascii_lowercase[:max(M_t.dim() - N_t.dim() - 1, 0)]
W_t = torch.einsum(prefix + "...yz,...z->" + prefix + "...y", [M_t, N_t])

torch.allclose(W_t, torch.from_numpy(W_np))  # True
```

**Comment by xvdp** [comment]
@Balandat in your tests, do measure memory under cuda while broadcasting, before, after, and even after cleanup and broadcast going out of scope. I dont have a good repro but I wonder if there may be leaks in some broadcasting mechanism. Then again i do understand some behaviour that looks like leaks is normal as cuda allocated memory is kept around, so  i dont know, but if you are working on that, do watch it.

**Comment by apaszke** [comment]
cc @t-vi

**Comment by t-vi** [comment]
Yeah, so we don't have "broadcasting" in einsum yet. It would probably not be too hard to do, but I must admit that it's not super-high on my priority list.
If anyone else wants to take a shot at implementing this and has specific questions, I'd be happy to help.

**Comment by Balandat** [comment]
@fritzo I know you guys use einsum a ton in pyro. How useful would this be to you?

**Comment by fritzo** [comment]
> How useful [bruadcasting via ellipsis notation] be to you?

Pyro does make heavy use of `torch.einsum` but we don't need ellipsis notation. Our highest priority request is for improved performance in simple equations with small dimensions :smile:

**Comment by t-vi** [comment]
@frizto CPU or GPU? I still have my TensoIterator `sumproduct_pair` collecting dust in some branch. It should not be too bad for GPU and I might add AVX to make it suck less on CPU.

**Comment by fritzo** [comment]
@t-vi CPU and GPU are both important for Pyro's use of `einsum`. I haven't profiled in the last 6 months but last time I did, torch CPU was still slower than numpy (even after #10661). Let me know if you start working on this, I can send you some example einsum inputs.
